### PR TITLE
Simplify multiplication by `0` and by `null` 

### LINKED ISSUE #3626: Simplify `A * 0` and `A * null`.
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

Add the following simplification rules:
-  `forall A:Expr: A * null --> null`
- `forall A:Expr, if A isn't nullable, A * 0 --> 0`

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>

# Which issue does this PR close?

Closes #3626.

 # Rationale for this change

Please look at the issue.

# What changes are included in this PR?

Add 2 simplification rules and tests

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 1]
```diff
@@ -110,6 +110,22 @@ fn expr_contains(expr: &Expr, needle: &Expr, search_op: Operator) -> bool {
...
+        | Expr::Literal(ScalarValue::Int16(Some(0)))
+        | Expr::Literal(ScalarValue::Int32(Some(0)))
+        | Expr::Literal(ScalarValue::Int64(Some(0)))
+        | Expr::Literal(ScalarValue::UInt8(Some(0)))
+        | Expr::Literal(ScalarValue::UInt16(Some(0)))
+        | Expr::Literal(ScalarValue::UInt32(Some(0)))
+        | Expr::Literal(ScalarValue::UInt64(Some(0))) => true,
+        Expr::Literal(ScalarValue::Float32(Some(v))) if *v == 0. => true,
+        Expr::Literal(ScalarValue::Float64(Some(v))) if *v == 0. => true,
+        _ => false,
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:125]: I wonder if we can also support `ScalarValue::Decimal`?

**Review by alamb — APPROVED** [review]
looks great -- thank you @HaoYang670

File: datafusion/optimizer/src/simplify_expressions.rs (hunk 1 above)
Comment (HaoYang670) [reply datafusion/optimizer/src/simplify_expressions.rs:125]: We could create a follow-up ticket to add Decimal.

**Comment by alamb** [comment]
Filed https://github.com/apache/arrow-datafusion/issues/3643 to track adding Decimal support

**Pull request merged**